### PR TITLE
update api for userproject with task and reflection

### PR DESCRIPTION
**Pull request by borkarsaish65**
# Description
These recommendations are intended to promote code quality and team communication during software development. They cover a variety of topics, including ensuring that pull requests are submitted to the correct branch, documenting new methods, preserving consistency across many services, and avoiding typical blunders like accessing APIs or DB queries within loops. Sensitive data should not be uploaded, and changes to environment variables or database models should be executed consistently. Teams may work more effectively and develop higher-quality software by adhering to these standards.

## Type of change
Please choose appropriate options.

- [ ]  Bug fix (non-breaking change which fixes an issue)
- [ ]  New feature (non-breaking change which adds functionality)
- [ ]  Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ]  Enhancement (additive changes to improve performance)
- [ ]  This change requires a documentation update

## Checklist 

- [ ]  It's critical to avoid making needless file modifications in contributions, such as adding new lines, console logs, or additional spaces, to guarantee cleaner and more efficient code. Furthermore, eliminating unnecessary imports from a file might enhance code readability and efficiency.
- [ ]  Ensure that the pull request is assigned to the right base branch and that the development branch name contains the JIRA Task Id. Furthermore, each commit message should include the JIRA Task Id in the manner "ED-100: message".
- [ ]  Only update packages if it is mentioned and authorized in the design document, and make sure that you have the required permissions.
- [ ]  Avoid making API and database queries inside a loop as it can lead to performance issues and slow down the system.
- [ ]  When calling another function inside a given function, add comments explaining the purpose and meaning of the passed arguments and expected return values.
- [ ]  If adding a blank argument in a function, add a comment explaining the reason for the blank argument.
- [ ]  Before submitting a pull request, do a self-review of your code to ensure there are no conflicts with the base branch and all comments have been addressed.
- [ ]  Before merging a pull request, it's important to have other team members review it to catch any potential errors or issues
- [ ]  To maintain code integrity, it's important to remove all related changes when removing code during a code review.
- [ ]  If new constants, endpoints, or utility functions are introduced, it is important to check if they already exist in the service to avoid any duplication.
- [ ]  Whenever a new environment variable is added to a service, it's important to ensure that the necessary changes are made to related files such as ".env.sample" and "envVariables.js" to maintain consistency and avoid errors. Additionally, the new environment variable should be added to the devops repository to ensure that it is properly documented and accessible to the team.
- [ ]  When adding a new function to a service, it is important to document it with relevant information such as the name, parameters, and return value in a consistent format across all services. Additionally, if there are any changes to the API response, ensure that the documentation in the controllers is updated accordingly. 
- [ ]  Write a clear and concise commit message that describes the changes made.
- [ ]  Maintain consistent function signature and code across all services when adding a function to multiple services. Implement changes to database models in all services that use the same model. 
- [ ]  Use only let and const. Do not use var.
- [ ]  Make common functions for repetitive code blocks.
- [ ]  Avoid uploading sensitive information such as secret tokens or passwords in pull requests to ensure data security.
- [ ]  Maintain consistent indentation and spacing throughout the code.

### CODE REVIEW

**Review by VISHNUDAS-tunerlabs — CHANGES_REQUESTED** [review]
reviewed 4 Dec

File: controllers/v1/userProjects.js [hunk 1]
```diff
@@ -1254,4 +1254,43 @@ module.exports = class UserProjects extends Abstract {
...
+            "_id": "667aabc7b070696248731e8e"
+        }
+    }
+	 * Verify project certificate
+	 * @method
+	 * @name update
+	 * @param {String} projectId - projectId.
+	 * @returns {JSON} certificate details.
+	 */
+	async update(req) {
```
[the remark this answers is not in the document]
Comment (borkarsaish65) [reply controllers/v1/userProjects.js:1277]: added

File: generics/constants/api-responses.js [hunk 2]
```diff
@@ -241,4 +241,6 @@ module.exports = {
 	WISHLIST_FETCH_FAILED: 'Wishlist fetch failed',
 	WISHLIST_ALREADY_ADDED: 'This MI is already added to the wishlist.',
 	PROJECT_TEMPLATE_NOT_IN_WISHLIST: 'This MI is not found in the wishlist.',
+	PROJECT_UPDATED_SUCCESSFULLY: 'Project Updated Successfully.',
+	PROJECT_UPDATED_FAILED: 'Project Updated Failed.',
```
[the remark this answers is not in the document]
Comment (borkarsaish65) [reply generics/constants/api-responses.js]: done

File: module/userProjects/helper.js [hunk 3]
```diff
@@ -3577,6 +3577,116 @@ module.exports = class UserProjectsHelper {
...
+	 * update project infromation
+	 * @method
+	 * @name update
+	 * @param {Object} filter - filter to search project to be updated.
+	 * @param {Object} updateTo - data which will be updated.
+	 * @param {String} userId - Logged in user Id.
+	 * @returns {Object} status of update project
+	 */
+
+	static update(filter, updateTo, userId) {
```
[the remark this answers is not in the document]
Comment (borkarsaish65) [reply module/userProjects/helper.js]: done

File: module/userProjects/helper.js [hunk 4]
```diff
@@ -3577,6 +3577,116 @@ module.exports = class UserProjectsHelper {
...
+
+				const userProject = await projectQueries.projectDocument(filter, ['_id', 'reflection', 'tasks'])
+
+				if (!(userProject && userProject[0])) {
+					throw new Error(CONSTANTS.apiResponses.USER_PROJECT_NOT_FOUND)
+				}
+
+				let currentReflection = userProject[0].reflection || {}
+
+				if (reflectionStatus == 'started') {
```
[the remark this answers is not in the document]
Comment (borkarsaish65) [reply module/userProjects/helper.js]: done

File: module/userProjects/helper.js [hunk 5]
```diff
@@ -3577,6 +3577,116 @@ module.exports = class UserProjectsHelper {
...
+	 * @param {Object} filter - filter to search project to be updated.
+	 * @param {Object} updateTo - data which will be updated.
+	 * @param {String} userId - Logged in user Id.
+	 * @returns {Object} status of update project
+	 */
+
+	static update(filter, updateTo, userId) {
+		return new Promise(async (resolve, reject) => {
+			try {
+				let { tasks, reflectionStatus, ...filteredUpdateData } = updateTo
```
[the remark this answers is not in the document]
Comment (borkarsaish65) [reply module/userProjects/helper.js]: refactored now

File: models/projects.js [hunk 6]
```diff
@@ -191,6 +191,18 @@ module.exports = {
 			type: Array,
 			default: [],
 		},
+		source: {
+			type: Object,
+			default: {},
+		},
+		reflection: {
+			type: Object,
+			default: {},
```
[the remark this answers is not in the document]
Comment (borkarsaish65) [reply models/projects.js]: done

File: controllers/v1/userProjects.js [hunk 7]
```diff
@@ -1254,4 +1254,50 @@ module.exports = class UserProjects extends Abstract {
...
+		return new Promise(async (resolve, reject) => {
+			try {
+				const updateData = await userProjectsHelper.update(
+					{ _id: req.params._id },
+					req.body,
+					req.userDetails.userInformation.userId,
+					req.userDetails.userToken,
```
Comment (VISHNUDAS-tunerlabs) [inline controllers/v1/userProjects.js:1284]: @borkarsaish65 user token is required?

File: module/userProjects/helper.js [hunk 8]
```diff
@@ -174,20 +184,60 @@ module.exports = class UserProjectsHelper {
...
+				if (!invokedViaUpdateApi && userProject[0].status == CONSTANTS.common.SUBMITTED_STATUS) {
 					throw {
 						status: HTTP_STATUS_CODE.bad_request.status,
 						message: CONSTANTS.apiResponses.FAILED_TO_SYNC_PROJECT_ALREADY_SUBMITTED,
 					}
 				}
 
+				let currentReflection = userProject[0].reflection || {}
+
+				if (data.reflectionStatus == 'started') {
```
Comment (VISHNUDAS-tunerlabs) [inline module/userProjects/helper.js:203]: @borkarsaish65 we might have constant defined for started and completed, please use them

File: module/userProjects/helper.js [hunk 9]
```diff
@@ -3577,6 +3627,57 @@ module.exports = class UserProjectsHelper {
...
+				if (updateResult.message == CONSTANTS.apiResponses.USER_PROJECT_UPDATED) {
+					return resolve({
+						message: CONSTANTS.apiResponses.USER_PROJECT_UPDATED,
+						success: true,
+						result: {
+							_id: filter._id,
+						},
+					})
+				} else {
+					throw new Error(CONSTANTS.apiResponses.PROJECT_UPDATE_FAILED)
```
Comment (VISHNUDAS-tunerlabs) [inline module/userProjects/helper.js:3669]: @borkarsaish65 Please throw error with proper message and status

File: module/userProjects/helper.js [hunk 10]
```diff
@@ -3894,6 +3995,21 @@ function _projectTask(tasks, isImportedFromLibrary = false, parentTaskId = '') {
 	return tasks
 }
 
+function _fillMissingTaskInformation(tasks, tasksFromDB) {
```
Comment (VISHNUDAS-tunerlabs) [inline module/userProjects/helper.js:3998]: @borkarsaish65 Please add function doc, and what will this function do?

**Review by VISHNUDAS-tunerlabs — CHANGES_REQUESTED** [review]
Reviewed on Dec 6

File: module/userProjects/helper.js (hunk 8 above)
Comment (borkarsaish65) [reply module/userProjects/helper.js:203]: done

File: module/userProjects/helper.js (hunk 9 above)
Comment (borkarsaish65) [reply module/userProjects/helper.js:3669]: done

File: module/userProjects/helper.js (hunk 10 above)
Comment (borkarsaish65) [reply module/userProjects/helper.js:3998]: done

File: controllers/v1/userProjects.js (hunk 7 above)
Comment (borkarsaish65) [reply controllers/v1/userProjects.js:1284]: required in the sync api

**Review by VISHNUDAS-tunerlabs — CHANGES_REQUESTED** [review]

File: module/userProjects/helper.js [hunk 11]
```diff
@@ -102,7 +102,16 @@ module.exports = class UserProjectsHelper {
...
-	static sync(projectId, lastDownloadedAt, data, userId, userToken, appName = '', appVersion = '') {
+	static sync(
+		projectId,
+		lastDownloadedAt,
+		data,
+		userId,
+		userToken,
+		appName = '',
+		appVersion = '',
+		invokedViaUpdateApi = false
```
[the remark this answers is not in the document]
Comment (borkarsaish65) [reply module/userProjects/helper.js]: done added in the function signature

**Review by VISHNUDAS-tunerlabs — APPROVED** [review]
Dec 6

**Pull request merged**